Handle termination signals during streaming

avikivity opened the pull request:
In b19792405faee91e, we changed the shutdown process not to rely on the global reactor-defined exit, but instead added a local variable to hold the shutdown state. However, we did not propagate that state everywhere, and now streaming processes are not able to abort.

Fix that by enhancing stop_signal with a sharded<abort_source> member that can be propagated to services. Propagate it to storage_service and thence to boot_strapper and range_streamer so that streaming processes can be aborted.

Fixes #4674
Fixes #4501 

Tests: unit (dev), manual bootstrap test

tgrabiec commented on dht/range_streamer.cc:
```diff
@@ -269,7 +269,8 @@ future<> range_streamer::stream_async() {
             } catch (...) {
                 logger.warn("{} failed to stream. Will retry in {} seconds ...", _description, sleep_time);
-                sleep_abortable(std::chrono::seconds(sleep_time)).get();
+                _abort_source.check();
```
doesn't sleep_abortable() already check this?

avikivity replied:
It does not, AFAICT.

tgrabiec replied:
`sleep_abortable()` constructs a `sleeper` object which calls `as.subscribe()` and if that returns a disengaged optional then it won't arm the timer but resolve with an exception instead. `abort_source::subscribe()` checks for `abort_requested()` and if so returns a disengaged optional.

So it seems that `sleep_abortable()` already aborts immediately when the abort_source is aborted, so calling `check()` before it seems not necessary.

avikivity replied:
Yes. I looked at it but misread it (and the behavior is expected). I'll remove the checks.

asias commented:
FYI, https://github.com/scylladb/scylla/issues/4674.

asias commented:
How does this series fix the "upgrade" issue #4167?

avikivity commented on main.cc:
```diff
@@ -84,30 +84,47 @@ void register_tracing_keyspace_backend(backend_registry& br);
         engine().handle_signal(SIGTERM, [this] { signaled(); });
     }
     ~stop_signal() {
+        _cond.broadcast();
```
Not needed. It is a remnant of a different mechanism to stop the loop that broadcasted aborts to all shards (in the end, I found that I had if (!_caught) { return } instead of if (_caught) { return }, and that the original broadcast mechanism was fine.

I'll remove this bit.

avikivity commented on main.cc:
```diff
@@ -84,30 +84,47 @@ void register_tracing_keyspace_backend(backend_registry& br);
     bool _caught = false;
     condition_variable _cond;
+    sharded<abort_source, plain_sharded_traits<abort_source>> _abort_sources;
+    //future<> _broadcasts_to_abort_sources_done = broadcast_loop();
```
yes (in fact the one I described earlier), will remove

avikivity commented:
I meant #4674, don't know how I typo'ed it so badly.

avikivity commented:
Updated. Changes
 - removed remnants of older cross-shard broadcast that was only partially removed during development
 - removed chatty log message leftover from development
 - corrected issue number in cover letter

bhalevy commented:
@espindola please review as well (remotely related to #4384)

avikivity commented:
Added reference to #4501 too

avikivity commented:
Updated:
 - remove redundant abort_source::check() when calling sleep_abortable() right after.

gleb-cloudius commented:
On Tue, Jul 09, 2019 at 11:27:10AM -0700, Avi Kivity wrote:
> Fix that by enhancing stop_signal with a sharded<abort_source> member that can be propagated to services. Propagate it to storage_proxy and thence to boot_strapper and range_streamer so that streaming processes can be aborted.
You mean storage_service, not storage_proxy, no?

--
			Gleb.

avikivity commented:
> On Tue, Jul 09, 2019 at 11:27:10AM -0700, Avi Kivity wrote: Fix that by enhancing stop_signal with a sharded<abort_source> member that can be propagated to services. Propagate it to storage_proxy and thence to boot_strapper and range_streamer so that streaming processes can be aborted.
> You mean storage_service, not storage_proxy, no?


Yes, updated.

espindola commented:
Benny Halevy <notifications@github.com> writes:

> @espindola please review as well (remotely related to #4384)

The general idea LGTM, I don't know the codebase well enough to say what
the best lines for adding calls to check() are.

Tests would be nice, if they are not too hard to do. Sending a signal to
self via a callback might do the trick, but if there is no handy
callback already, no need to add one just for testing.

Cheers,
Rafael

slivne commented:
@avikivity - is anything left ? if not @tgrabiec can you merge

tgrabiec commented:
I cannot merge it because it has this note:

> Note: do not merge, depends on unreviewed/unmerged seastar patch for plain_sharded_traits.

avikivity commented:
Updated. Changes:
 - rebased around seastar commit to accept sharded services without stop()
 - adjusted code to account for that (was using a version with plain_sharded_traits<>)
 - added missing std::ref() around sharded<abort_source> in tests/gossip.cc (not caught by tests since is isn't in test.py)
